bolt/lib/Passes/MarkRAStates.cpp

```diff
@@ -0,0 +1,122 @@
+#include "bolt/Core/ParallelUtilities.h"
+#include "bolt/Utils/CommandLineOpts.h"
+#include <cstdlib>
+#include <fstream>
+#include <iterator>
+
+#include <iostream>
+#include <optional>
+#include <stack>
+
+using namespace llvm;
+
+namespace llvm {
+namespace bolt {
+
+void MarkRAStates::runOnFunction(BinaryFunction &BF) {
+
+  if (BF.isIgnored())
+    return;
+
+  BinaryContext &BC = BF.getBinaryContext();
+
+  for (BinaryBasicBlock &BB : BF) {
+    for (auto It = BB.begin(); It != BB.end(); ++It) {
+      MCInst &Inst = *It;
+      if ((BC.MIB->isPSign(Inst) || BC.MIB->isPAuth(Inst)) &&
+          !BC.MIB->hasNegateRAState(Inst)) {
+        // no .cfi_negate_ra_state attached to signing or authenticating instr
+        // means, that this is a function with handwritten assembly, which might
+        // not respect Clang's conventions (e.g. tailcalls are always
+        // authenticated, so functions always start with unsigned RAState when
+        // working with compiler-generated code)
+        BF.setIgnored();
+        BC.outs() << "BOLT-INFO: ignoring RAStates in function "
+                  << BF.getPrintName() << "\n";
+        return;
+      }
+    }
+  }
+
+  bool RAState = false;
+  std::stack<bool> RAStateStack;
+
+  for (BinaryBasicBlock &BB : BF) {
+    for (auto It = BB.begin(); It != BB.end(); ++It) {
+
+      MCInst &Inst = *It;
+      if (BC.MIB->isCFI(Inst))
+        continue;
+
+      if (BC.MIB->isPSign(Inst)) {
+        assert(!RAState && "Signed RA State before PSign");
+        BC.MIB->setRASigning(Inst);
+
+      } else if (BC.MIB->isPAuth(Inst)) {
+        assert(RAState && "Unsigned RA State before PAuth");
+        BC.MIB->setAuthenticating(Inst);
+      } else if (RAState) {
+        BC.MIB->setRASigned(Inst);
+      } else {
+        BC.MIB->setRAUnsigned(Inst);
+      }
+
+      // Updating RAState. All updates are valid from the next instruction.
+      // Because the same instruction can have remember and restore, the order
+      // here is relevant. This is the reason to loop over Annotations instead
+      // of just checking each in a predefined order.
+      for (unsigned int Idx = 0; Idx < Inst.getNumOperands(); Idx++) {
+        std::optional<int64_t> Annotation =
+            BC.MIB->getAnnotationAtOpIndex(Inst, Idx);
+        if (!Annotation)
+          continue;
+        if (Annotation == MCPlus::MCAnnotation::kNegateState)
+          RAState = !RAState;
+        if (Annotation == MCPlus::MCAnnotation::kRememberState)
+          RAStateStack.push(RAState);
+        if (Annotation == MCPlus::MCAnnotation::kRestoreState) {
+          RAState = RAStateStack.top();
+          RAStateStack.pop();
+        }
+      }
+    }
+  }
+}
+
+Error MarkRAStates::runOnFunctions(BinaryContext &BC) {
+  ParallelUtilities::WorkFuncTy WorkFun = [&](BinaryFunction &BF) {
+    runOnFunction(BF);
+  };
+
+  ParallelUtilities::runOnEachFunction(
+      BC, ParallelUtilities::SchedulingPolicy::SP_TRIVIAL, WorkFun, nullptr,
```

Comment:
Is the policy chosen based on your discussion with @kbeyls  about costs?
Unsure what is more appropriate here. At least on theoretical time complexity you do a worst-case 2x linear pass over instructions.

Reply:
No, I just copied this from some other Pass I believe, and forgot to make a proper decision. Do you have a suggestion to which one should I use?

Reply:
I believe it should be `SP_INST_LINEAR` instead (same for `InsertNegateRAStatePass.cpp`).

Reply:
@maksfb , @yota9 can you comment on this?

Reply:
In [35306ef](https://github.com/llvm/llvm-project/pull/120064/commits/35306ef3cf370402717b8d1bd0ed8a829f0c4f09), I pushed some code related to the parallel execution, and wanted to fix this as well.

However, I've found that adding `SP_INST_LINEAR` breaks [`bolt/test/AArch64/unmarked-data.test`](https://github.com/llvm/llvm-project/blob/main/bolt/test/AArch64/unmarked-data.test). The test checks that no warnings are emitted, but we get this:

```
 // CHECK-NOT: BOLT-WARNING
               ^
<stdin>:7:1: note: found here
BOLT-WARNING: Running parallel work of 0 estimated cost, will switch to trivial scheduling.
```

Any tips on how to add `SP_INST_LINEAR` without breaking this?